src/azure-cli-core/azure/cli/core/_identity.py

```diff
@@ -134,9 +133,83 @@ def login_with_service_principal_certificate(self, client_id, certificate_path):
         credential = CertificateCredential(self.tenant_id, client_id, certificate_path, authority=self.authority)
         return credential
 
-    def login_with_msi(self):
-        # Use ManagedIdentityCredential
-        pass
+    MANAGED_IDENTITY_TENANT_ID = "tenant_id"
+    MANAGED_IDENTITY_CLIENT_ID = "client_id"
+    MANAGED_IDENTITY_OBJECT_ID = "object_id"
+    MANAGED_IDENTITY_RESOURCE_ID = "resource_id"
+    MANAGED_IDENTITY_SYSTEM_ASSIGNED = 'systemAssignedIdentity'
+    MANAGED_IDENTITY_USER_ASSIGNED = 'userAssignedIdentity'
+    MANAGED_IDENTITY_TYPE = 'type'
+    MANAGED_IDENTITY_ID_TYPE = "id_type"
+
+    def login_with_managed_identity(self, identity_id, resource):
+        from msrestazure.tools import is_valid_resource_id
+        from requests import HTTPError
+
+        credential = None
+        id_type = None
+        if identity_id:
+            # Try resource ID
+            if is_valid_resource_id(identity_id):
+                credential = ManagedIdentityCredential(resource=resource, msi_res_id=identity_id)
+                id_type = self.MANAGED_IDENTITY_RESOURCE_ID
+            else:
+                authenticated = False
+                try:
+                    # Try client ID
+                    credential = ManagedIdentityCredential(client_id=identity_id)
+                    id_type = self.MANAGED_IDENTITY_CLIENT_ID
+                    authenticated = True
+                except HTTPError as ex:
```

Comment:
call get_token before so that you can get the httperror?

Reply:
Let's wait until Azure Identity supports it. We don't know whether it will implement the failover logic.